src/Microsoft.Framework.Runtime.Sources/Impl/ApplicationGlobalData.cs

```diff
@@ -0,0 +1,84 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace Microsoft.Framework.Runtime.Sources.Impl
+{
+    // Helper class to completely centralize all of the application global data logic and avoid #if statements elsewhere.
+    internal class ApplicationGlobalData
+    {
+        private readonly IApplicationEnvironment _hostEnvironment;
+#if DNXCORE50
+        private readonly object _lock;
+        private readonly Dictionary<string, object> _appGlobalData;
+#endif
+
+        public ApplicationGlobalData(IApplicationEnvironment hostEnvironment)
+        {
+            _hostEnvironment = hostEnvironment;
+#if DNXCORE50
+            // If there is no host environment that we are delegating to, we need a place to store this data in CoreCLR
+            if (hostEnvironment == null)
+            {
+                // Initialize a new place to store global data on CoreCLR
+                _lock = new object();
+                _appGlobalData = new Dictionary<string, object>(StringComparer.Ordinal); // Case-sensitive, just like AppDomain.Get/SetData
+            }
+#endif
+        }
+
+        public object GetData(string name)
+        {
+            return _hostEnvironment == null ?
+                GetDataCore(name) :
+                _hostEnvironment.GetData(name);
+        }
+
+        public void SetData(string name, object value)
+        {
+            if (_hostEnvironment == null)
+            {
+                SetDataCore(name, value);
+            }
+            else
+            {
+                _hostEnvironment.SetData(name, value);
+            }
+        }
+
+#if DNX451
+        private object GetDataCore(string name)
+        {
+            return AppDomain.CurrentDomain.GetData(name);
+        }
+
+        private void SetDataCore(string name, object value)
+        {
+            AppDomain.CurrentDomain.SetData(name, value);
+        }
+#else
+        // NOTE(anurse): ConcurrentDictionary seems overkill here. This data is rarely used, and a global lock seems safer.
```

Comment:
It's overkill. This thing is rarely used. ConcurrentDictionary only makes sense in specific scenarios.